src/frontend/radeco_source.rs

```diff
@@ -0,0 +1,315 @@
+//! Defines the `Source` Trait.
+
+use serde_json;
+
+use std::path::{Path, PathBuf};
+use std::fs::{self, File};
+use std::io::{Read, Write};
+use std::rc::Rc;
+use std::cell::RefCell;
+use std::error::Error;
+use std::fmt;
+
+use r2api::api_trait::R2Api;
+use r2pipe::r2::R2;
+use r2api::structs::{FunctionInfo, LFlagInfo, LOpInfo, LRegInfo, LSectionInfo, LStringInfo, LSymbolInfo,
+LImportInfo, LExportInfo, LRelocInfo, LEntryInfo};
+
+#[derive(Debug)]
+pub enum SourceErr {
+    SrcErr(&'static str),
+    OtherErr(Box<Error>),
+}
+
+impl fmt::Display for SourceErr {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            &SourceErr::SrcErr(s) => write!(f, "{}", s),
+            &SourceErr::OtherErr(ref e) => write!(f, "{}", e),
+        }
+    }
+}
+
+impl<T: 'static + Error> From<T> for SourceErr {
+    fn from(e: T) -> SourceErr {
+        SourceErr::OtherErr(Box::new(e))
+    }
+}
+
+// TODO: Split this up/compose this from more basic traits to avoid reimplementation
+// as currently this is a re-implementation of r2api
```

Comment:
Yes, so if you've looked at the r2pipe-api [repo](https://github.com/radare/radare2-r2pipe-api/blob/master/rust/src/api_trait.rs#L13), I've separated (using comments) the different functions in r2api based on the functionality it provides (such as loader, arch, low-level analysis etc). These information can come from separate sources too. So imo, it makes sense to split them up into separate traits and have the current R2Api trait as a composition of these primitive traits:
`pub trait R2Api: `Loader + `Arch` ... { }`

Additionally, the current source takes `self` immutably, while r2api takes `self` mutably. This needs to be discussed. Perhaps we can reduce redundancy by implementing r2api over `Rc<RefCell<R2>>` (as it is currently here), rather than over R2 directly.

Reply:
I agree. Better to have the support of separate, independend sources.